File: Python/UnionFind.py
#Note: I am a C++ developer trying to teach myself Python, so my main focus is getting the code to work.

class UF:
    def __init__(self,N):
        self.id = []
        self.sz = []
        for i in range(N):
            self.id.append(i)
            self.sz.append(1)
    
    def connected(self, p, q):
        return self.root(p) == self.root(q)
    
    def root(self,i):
        while i != self.id[i]:
            self.id[i] = self.id[self.id[i]]
            i = self.id[i]
        return i
    
    def connect(self, p, q):
        i = self.root(p)
        j = self.root(q)
        if i == j:
            return
        if self.sz[i] > self.sz[j]:
            self.id[j] = self.id[i]
            self.sz[i] += self.sz[j]
        else:
            self.id[i] = self.id[j]
            self.sz[j] += self.id[i]

x = UF(10)
x.connect(1,3)
x.connect(3,4)
print(x.connected(1,4))
    

File: Java/UnionFind.java
public class UF {

    private int[] id;
    private int[] sz;

    public UF(int N){
        id = new int[N];
        sz = new int[N];

        for(int i = 0; i< N;i++)
        {
            id[i] = i;
            sz[i] = 1;
        }
    }
    private int root(int i )
    {
        while(i != id[i])
        {
            id[i] = id[id[i]];
            i = id[i];
        }
        return i;
    }
    public boolean connected(int p,int q)
    {
        return root(p) == root(q);
    }
    public void connect(int p, int q)
    {
        int i = root(p);
        int j = root(q);
        if (i == j) return;
        if (sz[i] > sz[j])
        {
            id[j] = i;
            sz[i] += sz[j];
        }
        else
        {
            id[i] = j;
            sz[j] += sz[i];
        }
    }
    public static void main(String[] args)
    {
        UF x = new UF(10);
        x.connect(1,4);
        x.connect(4,5);
        x.connect(5,8);
        System.out.println(x.connected(1,8));
    }

}
